include/bmi/Bmi_Py_Adapter.hpp

```diff
@@ -47,7 +47,8 @@ namespace models {
             Bmi_Py_Adapter(Bmi_Py_Adapter&&) = delete;
 
             ~Bmi_Py_Adapter() override {
-                Finalize();
+                if (bmi_model != NULL)
+                    Finalize();
```

Comment:
This is papering over something more fundamental - how are we ending up with a `Bmi_Py_Adapter` with a null `bmi_model`?  There's a class invariant problem here that's allowing that to happen.